src/dune_engine/load_rules.ml

```diff
@@ -32,7 +32,7 @@ module Loaded = struct
     }
 
   type t =
-    | Non_build of Path.Set.t
+    | Non_build of { files : Path.Set.t }
```

Comment:
Do I understand correctly that paths here can only be source or external? If yes, I think it's worth making the types more precise.

Reply:
Oh, wait, looks like they can only be source files, right? Then why isn't this `Path.Source.Set.t`?

Reply:
For example, we could have two constructors `Source` and `External`, with corresponding sets of paths, instead of one `Non_build`.

Reply:
> Do I understand correctly that paths here can only be source or external? 

Actually, it can only be external. In the test, i'm referring to source paths by an absolute path, so dune doesn't care where they actually exist and just treats them as external.

> If yes, I think it's worth making the types more precise.

Good point, but there's a bit of a trade-off here. It means that we'll need to convert it to `Path.Set.t` anyway when loading the rules. It seems like a bit of an extra conversion that doesn't pay off.

Reply:
> Actually, it can only be external.

Wow, even better!

I am confused. Why would the caller of `load_rules` want a less precise type? If you get a `Path.t` you have to accommodate for all possibilities, perhaps, filling out the corresponding branches with some `assert false`s. If you've got a `Path.External.t`,  then you don't need the `assert`s, and turning it to a `Path.t` is trivial: just append the `External` constructor.

Reply:
>  and turning it to a Path.t is trivial: just append the External constructor.

It is trivial for paths, but not for sets of paths! And that's the only use case we actually care about here.

By the way, it's not even obvious on how precise we should go. We could do:

```ocaml
| Non_build_external of Path.External.Set.t (* loading external paths *)
| Non_build_source of Path.Source.Set.t (* loading source paths *)
```

Or what you originally suggested:

```ocaml
| Non_build of { external : Path.External.Set.t ; source : Path.Source.Set.t }
```

The former is the most precise, but again this precision isn't being used anywhere.

So I propose we leave things as is and perhaps leave a comment for now.

Reply:
Of course, maybe I missed something in my quick look. If you really think that we can't afford to use the right API because of performance, let's use the above type but with `Path.Set.t`s instead and a comment explaining why we have to do that, and what needs to happen to avoid this (I guess, we need a better type for sets of paths?).

Reply:
(The reason I am so persistent about such a seemingly trivial matter is that I spent hours of debugging time figuring things out around the build systems engine and its interaction with loading rules, and I'm sure that having more precise APIs would help me a lot just to make sense of things. This is just one such spot but there are many! Pushing type precision slowly through the code base will pay back in saving development time, I'm sure.)

Reply:
Alright, I refactored it.